### DISCUSSION TITLE
How to access other attributes in the default function for setting an attribute's default value schematically?

### DISCUSSION THREAD

**Discussion by hdrhmd** [dynamoose/dynamoose#1480] [Q&A]
The example in the doc for using a function for setting the default value of an attribute doesn't take an argument. Is it possible to access other attributes?

I am looking for something like:
```
{
...
    "sk": {
        "type": String,
        "rangeKey": true,
        "default": (instance) => "someString#" + instance.someAttr
    },
...
}
```

**Comment by fishcharlie** [comment]
This is not currently supported. There are a lot of things to think through for this. For example, application of defaults happens pretty early in the Dynamoose process. Things like `get`/`set` modifiers haven't been applied yet, mappings fromDynamoDB haven't been applied yet, Combine properties haven't been applied yet.

The closest thing that we have currently is a `set` function on the Schema settings. The downside to that is that it gets called for everything. So even `get`, `query`, `update` calls will use that. I doubt that is what you are looking for.

I'm curious to hear more about your use case for this. What exactly is `someAttr`? What are you trying to achieve here? And why do you want this functionality?

**Reply by hdrhmd** [reply]
Thanks for responding. The use case that I am trying to implement is related to overloading global secondary indexes when using the single table design. To ensure we are able to reuse/overload the same global secondary index for multiple entities and scenarios we have to schematically set values on these indexes.

For example, let's say we have a many-to-many relationship between a User entity and a Course entity. The associative entity that captures the relationship is the UserCourse entity shown below.

```
class UserCourse extends Item {
  userId: number;
  courseId: number;
  courseName: string;
  completedAt?: string;
  ...
}
```

For the UserCourse entity I want to set GSI3-PK and GSI3-SK as follows:
```
"GSI3-PK": (userCourse) => "user#" + userCourse.userId;
"GSI3-SK": (userCourse) => userCourse.completedAt ? "course-completedAt#" + userCourse.completedAt : "course";
```

Example values:
```
GSI3-PK = "user#12345"
GSI3-SK = "course" // when course is not completed yet
GSI3-SK = "course-completedAt#2022:02:01T03:12:00Z" // when course was completed at 2022:02:01T03:12:00Z
```

By schematically setting values for the global index in this example GSI3, I am able to reuse it for all entities and query it for different scenarios, for example:
1. To join all entities related to the user: Query(GSI3-PK=user#12345)
2. To get all courses of a user: Query(GSI3-PK=user#12345 AND GSI3-SK begins_with(course)) 
3. To get all courses of a user that are not completed yet: Query(GSI3-PK=user#12345 AND GSI3-SK=course)
4. To get all courses of a user that have been completed: Query(GSI3-PK=user#12345 AND GSI3-SK begins_with(course-completedAt#))

This video from AWS re:invent 2021 captures advanced design patterns that will benefit from being able to schematically set values on indexes: https://www.youtube.com/watch?v=xfxBhvGpoa0

**Reply by hdrhmd** [reply]
I found a better example on the internet that explains my use case better than my poorly made-up example: https://www.sensedeep.com/blog/posts/2021/dynamodb-singletable-design.html

In the example below you can see how the pk and sk values are schematically set. To be able to do that, we need access to other attributes on the model which is what my original question was about.

```
const MySchema = {
    indexes: { primary: { hash: 'pk', sort: 'sk' } },
    models: {
        Account: {
            pk:          { value: 'account#${name}' },
            sk:          { value: 'account#' },
            name:        { type: String, required: true },
        },
        User: {
            pk:          { value: 'account#${accountName}' },
            sk:          { value: 'user#${email}', validate: EmailRegExp },
            accountName: { type: String },
            email:       { type: String, required: true },
        },
        Post: {
            pk:          { value: 'post#${id}' },
            sk:          { value: 'user#${email}', validate: EmailRegExp },
            id:          { type: String },
            message:     { type: String },
            email:       { type: String },
        }
    },
    version: '0.1.0',
    format: 'onetable:1.0.0'
}
```

**Reply by fishcharlie** [reply]
@hdrhmd Sorry for the delay here. I had a reply written yesterday, and GitHub deleted it when I closed out of the tab by accident. **SUPER** frustrating.

Anyways. This is a perfect use case for the `Combine` Dynamoose property type. In fact this is the exact use case for why the `Combine` Dynamoose property type was created.

Let's look at your example here:

```js
Account: {
    pk:          { value: 'account#${name}' },
    sk:          { value: 'account#' },
    name:        { type: String, required: true },
}
```

You can extend this to your specific use case, but this one is simple enough to get started with. I'm also going to ignore the `sk` part and focus on the `pk` & `name` attributes here, just for sake of simplicity.

So the `pk` attribute if we break it apart is a static type string, and the value of the `name` attribute, separated by a `#` character. So the first thing we want to do is create a [`Constant` property type](https://dynamoosejs.com/guide/Dynamoose#dynamoosetypeconstantvalue) in Dynamoose. In your schema that might look something like this:

```js
"type": dynamoose.type.CONSTANT("account")
```

Then, we can set the `pk` atttribute to be type [`Combine`](https://dynamoosejs.com/guide/Dynamoose#dynamoosetypecombineattributes-separator).

```js
"pk": {
	"type": dynamoose.type.COMBINE(["type", "name"], "#")
}
```

Or

```js
"pk": dynamoose.type.COMBINE(["type", "name"], "#")
```

(Both the same thing, just wanted to demonstrate that both options would work)

There are few benefits to this. With a `default` value, if you ever updated the `name` property, you would have to set `pk` to `undefined` to get it to revert to the default value. This increases the risk of forgetting to do that, and having a mismatch between the `pk` and `name` attributes. With the `Combine` property type, it will always be updated whenever you update either of thsoe attributes. This is probably an extremely rare use case, and will never happen with a `pk` attribute, but it is a use case that Dynamoose helps guard against that a `default` value would not.

Only downside to this is that you are now saving the `type` attribute in DynamoDB as well. Maybe in the future it'd be cool to have an option to have a Dynamoose attribute that doesn't get saved to DynamoDB??? If you think that'd be useful, feel free to submit an issue on this repository as a Feature Suggestion.

I'd also encourage you to look at the documentation for [`map`](https://dynamoosejs.com/guide/Schema#map-string--string). [`Combine`](https://dynamoosejs.com/guide/Dynamoose#dynamoosetypecombineattributes-separator), [`Constant`](https://dynamoosejs.com/guide/Dynamoose#dynamoosetypeconstantvalue) & [`map`](https://dynamoosejs.com/guide/Schema#map-string--string) are the most important Dynamoose concepts when using single table design.

---

I hope this helps. Let me know if you have any questions.

**Reply by hdrhmd** [reply]
Hey, @fishcharlie apologies for a very late reply. And thank you for taking the time to reply to my question in such detail (and that too twice)! I have been looking at other libraries to see how they handle my use case. DynamoDB Toolbox supports the kind of [custom function](https://www.dynamodbtoolbox.com/docs/entity#customize-defaults-with-a-function) that I had asked for in my original post. It also has a feature similar to the [combine feature](https://www.dynamodbtoolbox.com/docs/entity#using-an-array-for-composite-keys), though their documentation discourages its use.

For my use case, I needed the flexibility provided by the custom function for setting defaults so I went with DynamoDB Toolbox. I recommend considering implementing a custom function for defaults in Dynamoose as it gives greater flexibility required for schematic index values. I am not sure how `combine` will handle cases where we want to set different values based on some condition. Custom functions can do that.

Example of using condition in setting the value:
```
"GSI3-SK": (userCourse) => userCourse.completedAt ? "course-completedAt#" + userCourse.completedAt : "course";
```
